keps/20200225-reading-parameter-values-from-a-file.md

```diff
@@ -0,0 +1,120 @@
+---
+kep-number: draft-20200225
+title: Reading parameter values from a file
+short-desc: Reading parameter values from a file
+authors:
+  - "@porridge"
+owners:
+  - "@porridge"
+editor: @porridge
+creation-date: 2020-02-25
+last-updated: 2020-02-25
+status: provisional
+see-also:
+  - KEP-24
+  - KEP-25
+---
+
+# Reading parameter values from a file
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP describes an alternative way to provide parameter values when installing or
+updating an operator instance.
+
+## Motivation
+
+This has two motivations:
+1. A better way for a user to store parameter values than a shell command line.
+1. A more convenient way to pass complex parameter values, especially in the light of
+coming structured parameter values (see [KEP-24](https://github.com/kudobuilder/kudo/pull/1356)).
+
+See [User Stories](#user-stories) below for more details.
+
+### Goals
+
+- Make it possible to store parameter values in a file, and pass them in this form
+to any `kubectl kudo` command which currently takes the `-p` / `--parameter` flag.
+- Support both currently supported (i.e. `string`) and planned (see [KEP-24](https://github.com/kudobuilder/kudo/pull/1356))
+types of parameter values.
+
+### Non-Goals
+
+- Make it possible for a user to pass any other command line flags through the file.
+- Provide any templating or other pre-processing capabilities for the parameter value file.
+
+## Proposal
+
+The idea is to introduce a new command-line flag: `--parameter-file`, in every place that
+currently accepts the `--parameter` flag.
+This flag should require a single argument, interpreted as a path to a YAML file.
```

Comment:
I think I would prefer that this is **NOT** a yaml file.   Is there a reason to support YAML here?

My preference is to use properties files... this is already familiar to k8s users which using configmap configurations via files:  https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/
example file referenced is game.properties which is https://kubernetes.io/examples/configmap/game.properties

Ours should be <param>=<value> 

What we need to understand is... do we error on unknown?

Reply:
Is there a reason *not* to use a a yaml?

For simple parameter values (ints, strings, etc.) a properties file would work fine, but as we're going to add more complex parameter values (maps, lists), yaml would provide a good structure. I think that's where the KEP-25 comes into play.


What do you mean with `error on unknown`?

Reply:
through meetings it has become clear the need / value of yaml (which is captured by ANeumann82 ^^.   I believe we should allow for both approaches.  I think properties files are simple, well understood, used in the eco-system and capture the 80+% use cases.

The question becomes... do we have 2 flags (not my preference) or do we detect the file type.   We could simply use the file extension as the file detection mechanism.

Reply:
Thanks, @kensipe. My two cents here: I think that `.properties` files are an acceptable solution for passing configuration to applications (via ConfigMap), however, given that `operator.yaml`, `params.yaml`, and the majority of templates are using YAML format, `.properties` files look a bit alien in this environment and introduce yet another format in addition to YAML-centric approach used for defining K8s resources.

Let's say a user started with `.properties` file and then at some point decided to add some complex parameter value. `.properties` file stops working and the user needs to migrate to YAML. The configuration codebase starts diverging with some operators configured with `.properties` and some in YAML - I don't think that supporting both formats brings any benefits in this scenario. Having a single format which is not limited when it comes to full parameter support is a better option here IMO.

Reply:
Funny how @akirillov expressed my own thoughts on this topic better than I would have :-)

Reply:
@akirillov sold me!  @porridge I do feel that this kep should make an attempt at defining the yaml schema... or an example layout.

Reply:
Added a section on schema: 41549a4c.